session/temptable/manager.go

```diff
@@ -0,0 +1,312 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package temptable
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/pingcap/tidb/util/tableutil"
+
+	"github.com/pingcap/parser/model"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/sessionctx"
+	transaction "github.com/pingcap/tidb/store/driver/txn"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	tikvkey "github.com/tikv/client-go/v2/kv"
+	"github.com/tikv/client-go/v2/tikv"
+)
+
+// SessionTemporaryTableManager is used to manage temporary tables
+type SessionTemporaryTableManager struct {
+	ctx         sessionctx.Context
+	localTables *infoschema.LocalTemporaryTables
+	memData     *tikv.MemDB
+
+	cache struct {
+		schemaMetaVersion    int64
+		sortedGlobalTables   []table.Table
+		sortedTableKeyRanges []*tikv.RangeRetriever
+	}
+}
+
+// GetTemporaryTableManager get the TemporaryTableManager in session
+func GetTemporaryTableManager(ctx sessionctx.Context) *SessionTemporaryTableManager {
+	sessionVars := ctx.GetSessionVars()
+	mgr := sessionVars.TemporaryTableManager
+	if mgr == nil {
+		mgr = &SessionTemporaryTableManager{ctx: ctx}
+		sessionVars.TemporaryTableManager = mgr
+	}
+
+	return mgr.(*SessionTemporaryTableManager)
+}
+
+// UpdateSnapshotOptions is used for update snapshot options when schema changed
+func (m *SessionTemporaryTableManager) UpdateSnapshotOptions(snap kv.Snapshot, is infoschema.InfoSchema) {
+	if is.SchemaMetaVersion() != m.cache.schemaMetaVersion {
```

Comment:
I have some concerns.
1. Does the schema version change when you only create / drop / truncate a local temporary table?
2. I'm afraid the performance may jitter in some extreme situations. E.g. many temporary tables, or frequent DDLs.

Reply:
1. schema version will not change when create / drop / truncate a local temporary table. But the cache will still be rebuilt to make all things right.
2. The performance concern for ddl changes and too many temporary tables should be discussed separately.
    * `sortedGlobalTemporaryTables` will be stored in each tidb instance for every schema version. That is, when creating/droping/truncate a global temporary table we will use `sort.Sort` to resort all temporary tables. But I don't think it is a big problem because 1. sort operation is executed in another goroutine, it'll not block other sql operations. 2. I wonder wether user should have more than 1000 temporary tables, if table count is not very big, cpu cost can be ignored even for sorting. 3. sort operation can be optimized because table id always increase，so if we found the new table id is bigger than any other table ids, we can just append it to the array.
    * We also store an array of sorted local temporary tables in every session. Every time ddl changes for local temporary table it also should be rebuilt. It is not a problem because it can also be optimized just global temporary table. The only one problem is, when creating a new transaction, we should make a merge sort for local and global temporary tables with time complexity O(m+n), m,n stands for count of local and global temporary tables. But I think it is acceptable and we have a cache to cache the result.
    * The biggest problem is that if we have many temporary tables. It may cost some cpu cycles to find the correct range the key located. Here I used a binary search to reduce the cpu cost(That is why we should sort all temporary tables). If we have 1000+ temporary tables, there are at most 10 compares. I think it takes much less time than a network round trip.

Reply:
More optimization can be done for data query. For example, if we found all income ranges is in prefix mod (KeyStart and KeyEnd have the some prefix unless the last byte and KeyEnd[len - 1] = keyStart[len + 1]), we can save these key prefixes into a map. O(1) can be archived for point get and scan query whose start and end have a same prefix

Reply:
Sorry, I didn't notice that it's a merge sort. I think it's fine.